### PR TITLE
PIM-10192: Retry mechanism in  delete  action of ES documents

### PR DESCRIPTION
**Pull request by rfauglas**
**Description (for Contributor and Core Developer)**
https://akeneo.atlassian.net/browse/PIM-10192
After analysis it is an internal Error caused by ElasticSearch due to concurrent access on ressources protected by optimistic lock.
I chose to implement in the Client Adaptation layer.

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [x] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php [hunk 1]
```diff
@@ -378,13 +386,27 @@ public function getConfigurationLoader(): Loader
...
+            $attempts++;
+            try {
+                $this->client->deleteByQuery([
+                    'index' => $this->indexName,
+                    'body' => $body,
+                ]);
+                return;
+            } catch (Conflict409Exception $e) {
+                $exception = $e;
+                sleep($this->maxExpectedIndexationLatency);
```
Comment (jmleroux) [inline src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php:405]:
```suggestion
                usleep($this->maxExpectedIndexationLatency);
```
Sleep parameter is in seconds. 100 seconds is huge. I guess you wanted 100 microsecs.

Comment (rfauglas) [reply src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php]: I convert  it in constructor... I wanted milliseconds...

**Review by RiffVentura — APPROVED** [review]

**Review by nmarniesse — APPROVED** [review]

File: src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php [hunk 2]
```diff
@@ -32,6 +33,9 @@ class Client
     private NativeClient $client;
     private string $idPrefix;
     private int $maxChunkSize;
+    private int $maxExpectedIndexationLatency;
```
Comment (ahocquard) [inline src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php:36]:
```suggestion
    private int $maxExpectedIndexationLatencyInMs;
```

Always better to specify the unit in the name, it often avoids many issues.

**Review by ahocquard — APPROVED** [review]
Pretty clean to do that in the client :+1:

File: src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php (hunk 2 above)
Comment (nmarniesse) [reply src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php:36]: I hesitate to add this comment too :)  
But `Ms` is usually milliseconds, not microseconds

Comment (ahocquard) [reply src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php:36]: I thought it was milliseconds, that's why. hence the importance to specify it.
`maxExpectedIndexationLatencyInMicroseconds`. Long but explicit.

Comment (rfauglas) [reply src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php]: i agree with both of you I'd add  we miss type alias to express units, which can be got around with naming conventions...
configuration in units can be precised in configuration. See my last proposal in last commit.

Comment (ahocquard) [reply src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php]: I would change the key configuration also: `max_expected_indexation_latency`

**Pull request merged**